# Ordinamento per fusione (merge sort)

Ripreniamo un problema che abbiamo già affrontato nella lezione del 23 novembre: data una lista `l`, vogliamo riordinare i suoi elementi dal più piccolo al più grande. Abbiamo già visto un algoritmo che risolve questo problema, il *selection sort* (ordinamento per selezione).

L'algoritmo di ordinamento per fusione (merge sort) si basa sull'idea di dividere a metà la lista, ordinare sepratamente le due parti e rifonderle (ordinate) nella lista di partenza. Il caso base è costitutito dalle liste di lunghezza 0 oppure 1, che sono sempre ordinate.

In [49]:
def merge_sort(l):
    """
    Ordina gli elementi di l dal più piccolo al più grande.
    """
    # CASO BASE: se l ha lunghezza 0 oppure 1, è già ordinata, e usciamo subito.
    if len(l) <= 1: return
    # CASO RICORSIVO
    # separo la lista l in due metà, first e second
    mid = len(l) // 2
    first = l[0:mid]
    second = l[mid:len(l)]
    # richiamo ricorsivamente merge_sort sulle due sottoliste
    merge_sort(first)
    merge_sort(second)
    # metto assieme i risultati
    merge(first, second, l)

La funzione `merge_sort` si basa sulla funzione `merge`, che mette assieme due liste precedentemente ordinate.

In [51]:
def merge(l1, l2, l):
    """
    Date due liste ordinate l1 ed l2, copia i loro elementi nella lista l
    preservando l'ordine complessivo. Per ipotesi, la lunghezza di l è
    almeno pari alla somma delle lunghezze di l1 ed l2.
    """
    i1 = 0   # indice dell'elemento da osservare in l1
    i2 = 0   # indice dell'elemento da osservare in l2
    idst = 0 # indiece dell'elemnento da sovrascrivere in l
    while i1 < len(l1) and i2 < len(l2):
        if l1[i1] < l2[i2]:
            l[idst] = l1[i1]
            i1 += 1
            idst += 1
        else:
            l[idst] = l2[i2]
            i2 += 1
            idst += 1

    # Copia la parte rimanente della lista l1 dentro l
    while i1 < len(l1):
        l[idst] = l1[i1]
        i1 += 1
        idst += 1

    # Copia la parte rimanente della lista l2 dentro l
    while i2 < len(l2):
        l[idst] = l2[i2]
        i2 += 1
        idst += 1

Vediamo come usare `merge` per fondere le liste `[1, 10]` e `[2, 4]`. Si noti che in questa implementazione bisogna fornire come terzo parametro di `merge_sort` la lista destinazione che deve avere già una lunghezza sufficiente per contenere il risultato. La cosa è un po' strana in generale, ma è la versione più comoda da usare per la funzione `merge_sort`.

In [54]:
l = [0] * 4
merge([1, 10], [2, 4], l)
l

[1, 2, 4, 10]

E questo è un esempio di utilizzo di `merge_sort`.

In [56]:
l = [15, 20, 12, -2, 75, 4, 6, 1]
merge_sort(l)
l

[-2, 1, 4, 6, 12, 15, 20, 75]

Ricordiamo per completezza il codice della funzione `selection_sort`.

In [57]:
def minimum_position(l, start):
    """
    Restituisce la posizione dell'elemento minimo nella sottolista l[start:].
    Si assume che start sia una posizione valida per l.
    """
    minpos = start
    for i in range(start + 1, len(l)):
        if l[i] < l[minpos]:
            minpos = i
    return minpos

def selection_sort(l):
    """Ordina gli elementi di l dal più piccolo al più grande."""
    for i in range(len(l)-1):
        m = minimum_position(l, i)
        l[i], l[m] = l[m], l[i]

# Tempo di esecuzione

## Liste piccole

Il codice di `merge_sort` è più complesso di quello di `selection_sort`. La cosa sembra ripercuotersi nel tempo di esecuzione, almeno per le liste piccole.

In [ ]:
%%timeit
l = [15, 20, 12, -2, 75, 4, 6, 1]
selection_sort(l)

1.87 µs ± 114 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [36]:
%%timeit
l = [15, 20, 12, -2, 75, 4, 6, 1]
merge_sort(l)

4.6 µs ± 261 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


Si vede che il merge sort per questa lista di esempio è circa 3 volte più lento del selection sort.

## Liste un po' più grandi

Riutilizziamo la funzione `create_random_list(n)` già vista nella lezione del 23 novembre.

In [58]:
from random import randint

def create_random_list(n):
    """
    Restituisce una lista di lunghezza `n` riempita di numeri interi casuali tra
    zero e un milione.
    """
    l = []
    for _ in range(n):
        l.append(randint(0, 1_000_000))
    return l

In [59]:
create_random_list(20)

[634626,
 297055,
 880161,
 825470,
 452972,
 873262,
 978584,
 258307,
 666917,
 325511,
 891502,
 314571,
 77259,
 44754,
 187838,
 736296,
 173731,
 892838,
 103180,
 276672]

In [61]:
l_lunga = create_random_list(50)

In [64]:
%%timeit
# Notare l'uso di "l = l_lunga[:]" invece di "l = l_lunga" perché voglio che l sia una copia
# di l_lunga, e non un alias. Infatti, non voglio che ordinare l abbia come effetto collaterale
# quello di ordinare anche l_lunga.
l = l_lunga[:]
selection_sort(l)

29.8 µs ± 969 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [65]:
%%timeit
l = l_lunga[:]
merge_sort(l)

38.4 µs ± 838 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


Notare che il selection sort è ancora più veloce del merge sort (almeno sul mio computer) ma il margine di miglioramento si è ridotto.

Proviamo con una lista ancora più grande.

In [67]:
l_più_lunga = create_random_list(100)

In [69]:
%%timeit
l = l_più_lunga[:]
selection_sort(l)

99.2 µs ± 332 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [70]:
%%timeit
l = l_più_lunga[:]
merge_sort(l)

85.3 µs ± 2.02 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


Adesso il merge sort ha preso il sopravvento e risulta più veloce del selection sort.

## Liste molto grandi

Riprendiamo un pezzo di codice scritto per la lezione del 23 novembre che prova ad ordine liste di lunghezza crescente da 1000 fino a 15.000 elementi.

In [71]:
from time import time

times = []
for i in range(1000, 15001, 1000):
    l = create_random_list(i)
    # La funzione time restituisce il numero di secondi trascorsi dal 1° gennaio 1970.
    start_time = time()
    selection_sort(l)
    end_time = time()
    # La differenza tra end_time e start_time è la quantitià di tempo, in secondi, che è
    # stata impiegata dalla funzione selection_sort.
    times.append(end_time - start_time)
    print(f"Size {i} Selection Sort Time: {end_time - start_time:10.3f} secondi")

Size 1000 Selection Sort Time:      0.016 secondi
Size 2000 Selection Sort Time:      0.064 secondi
Size 3000 Selection Sort Time:      0.142 secondi
Size 4000 Selection Sort Time:      0.254 secondi
Size 5000 Selection Sort Time:      0.399 secondi
Size 6000 Selection Sort Time:      0.569 secondi
Size 7000 Selection Sort Time:      0.812 secondi
Size 8000 Selection Sort Time:      1.052 secondi
Size 9000 Selection Sort Time:      1.273 secondi
Size 10000 Selection Sort Time:      1.588 secondi
Size 11000 Selection Sort Time:      2.009 secondi
Size 12000 Selection Sort Time:      2.294 secondi
Size 13000 Selection Sort Time:      2.897 secondi
Size 14000 Selection Sort Time:      3.248 secondi
Size 15000 Selection Sort Time:      3.595 secondi


In [72]:
times = []
for i in range(1000, 15001, 1000):
    l = create_random_list(i)
    # La funzione time restituisce il numero di secondi trascorsi dal 1° gennaio 1970.
    start_time = time()
    merge_sort(l)
    end_time = time()
    # La differenza tra end_time e start_time è la quantitià di tempo, in secondi, che è
    # stata impiegata dalla funzione selection_sort.
    times.append(end_time - start_time)
    print(f"Size {i} Merge Sort Time: {end_time - start_time:10.3f} secondi")

Size 1000 Merge Sort Time:      0.001 secondi
Size 2000 Merge Sort Time:      0.006 secondi
Size 3000 Merge Sort Time:      0.005 secondi
Size 4000 Merge Sort Time:      0.008 secondi
Size 5000 Merge Sort Time:      0.008 secondi
Size 6000 Merge Sort Time:      0.010 secondi
Size 7000 Merge Sort Time:      0.025 secondi
Size 8000 Merge Sort Time:      0.013 secondi
Size 9000 Merge Sort Time:      0.015 secondi
Size 10000 Merge Sort Time:      0.017 secondi
Size 11000 Merge Sort Time:      0.019 secondi
Size 12000 Merge Sort Time:      0.021 secondi
Size 13000 Merge Sort Time:      0.023 secondi
Size 14000 Merge Sort Time:      0.025 secondi
Size 15000 Merge Sort Time:      0.042 secondi


Si vede che il merge sort per liste così lunghe è deisamente più veloce del selection sort. Evidentemente il selection sort è più efficiente per liste corte, ma il merge sort è più veloce per liste lunghe.

Notiamo anche che quando raddoppiamo la dimensione della lista, il tempo di esecuzione del selection sort circa si quadruplica (nel passaggio da 5000 a 10.000 elementi il tempo passa da 0.399 a 1.588 secondi) mentre quello del merge sort raddoppia (o poco più). Abbiamo già visto infatti che il selection sort ha un tempo di esecuzione dell'ordine di $O(n^2)$: se $n$ raddoppia, $n^2$ quadruplica. Una analisi teorica ci mostrerà che il tempo di esecuzione del merge sort è invece dell'ordine di $O(n \log n)$, che è poso di più di una funzione linerare $O(n)$.